README.md
# sl-phone-number

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

Simple package to validate Sri Lankan phone numbers.

## Structure

If any of the following are applicable to your project, then the directory structure should follow industry best practices by being named the following.

```
bin/        
build/
docs/
config/
src/
tests/
vendor/
```


## Install

Via Composer

``` bash
$ composer require apichef/sl-phone-number
```

## Usage

``` php
$skeleton = new APiChef\SlPhoneNumber();
echo $skeleton->echoPhrase('Hello, League!');
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

``` bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md) for details.

## Security

If you discover any security related issues, please email milroy@outlook.com instead of using the issue tracker.

## Credits

- [Milroy .E Fraser][link-author]
- [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/apichef/sl-phone-number.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/apichef/sl-phone-number/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/apichef/sl-phone-number.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/apichef/sl-phone-number.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/apichef/sl-phone-number.svg?style=flat-square

[link-packagist]: https://packagist.org/packages/apichef/sl-phone-number
[link-travis]: https://travis-ci.org/apichef/sl-phone-number
[link-scrutinizer]: https://scrutinizer-ci.com/g/apichef/sl-phone-number/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/apichef/sl-phone-number
[link-downloads]: https://packagist.org/packages/apichef/sl-phone-number
[link-author]: https://github.com/milroyfraser
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All notable changes to `sl-phone-number` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## [0.1.2] - 2017-08-30
### Added
- toLocalFormat and toIDDFormat [#2](https://github.com/apichef/sl-phone-number/issues/2)


tests/ExampleTest.php
<?php

declare(strict_types=1);

namespace APiChef\SlPhoneNumber;

class ExampleTest extends \PHPUnit\Framework\TestCase
{
    /**
     * Test that true does in fact equal true
     */
    public function testTrueIsTrue()
    {
        $this->assertTrue(true);
    }
}


tests/PhoneNumberTest.php
<?php

declare(strict_types=1);

namespace APiChef\SlPhoneNumber;

use PHPUnit\Framework\TestCase;

class PhoneNumberTest extends TestCase
{
    /**
     * @dataProvider phoneNumbersWithValidity
     */
    public function test_getDetails_throws_an_exception_when_the_starting_digits_or_length_is_invalid(string $number, bool $validity, string $formated = null)
    {
        $slPhoneNumber = new PhoneNumber($number);

        if (! $validity) {
            $this->expectException(\InvalidArgumentException::class);
        }

        /** @var PhoneNumberDetails $result */
        $result = $slPhoneNumber->getDetails();

        $this->assertEquals($formated, $result->getNumber());
    }

    /**
     * @dataProvider phoneNumbersWithValidity
     */
    public function test_it_validates_phone_number_by_starting_digits_and_length(string $number, bool $validity)
    {
        $slPhoneNumber = new PhoneNumber($number);

        $this->assertEquals($validity, $slPhoneNumber->isValid());
    }

    public function phoneNumbersWithValidity()
    {
        return [
            ['0770000000', true, '0770000000'], // 10 digits
            ['0112000000', true, '0112000000'], // 10 digits
            ['1000000000', false, null], // 10 digits NOT starting with 0
            ['000000000', false, null], // 9 digits
            ['00000000000', false, null], // 11 digits

            ['+94770000000', true, '0770000000'], // +94 and 9 digits
            ['+94112000000', true, '0112000000'], // +94 and 9 digits
            ['-94000000000', false, null], // -94 and 9 digits
            ['+9400000000', false, null], // +94 and 8 digits
            ['+940000000000', false, null], // +94 and 10 digits

            ['0094770000000', true, '0770000000'], // 0094 and 9 digits
            ['0094112000000', true, '0112000000'], // 0094 and 9 digits
            ['0096000000000', false, null], // 0096 and 9 digits
            ['009400000000', false, null], // 0094 and 8 digits
            ['00940000000000', false, null], // 0094 and 10 digits
        ];
    }

    public function test_get_details_returns_phone_number_details_object()
    {
        $details = (new PhoneNumber('0710000000'))->getDetails();

        $this->assertInstanceOf(PhoneNumberDetails::class, $details);
        $this->assertEquals('Mobitel', $details->getOperator());

        $fixedDetails = (new PhoneNumber('0112000000'))->getDetails();

        $this->assertInstanceOf(PhoneNumberDetails::class, $fixedDetails);
        $this->assertEquals('Sri Lanka Telecom', $fixedDetails->getOperator());
        $this->assertEquals('Western', $fixedDetails->getProvince());
        $this->assertEquals('Colombo', $fixedDetails->getDistrict());
        $this->assertEquals('Colombo', $fixedDetails->getArea());
    }

    public function test_it_throws_a_logic_exception_when_trying_to_get_province_from_mobile_number()
    {
        $details = (new PhoneNumber('0710000000'))->getDetails();
        $this->expectException(\LogicException::class);
        $details->getProvince();
    }

    public function test_it_throws_a_logic_exception_when_trying_to_get_district_from_mobile_number()
    {
        $details = (new PhoneNumber('0710000000'))->getDetails();
        $this->expectException(\LogicException::class);
        $details->getDistrict();
    }

    public function test_it_throws_a_logic_exception_when_trying_to_get_area_from_mobile_number()
    {
        $details = (new PhoneNumber('0710000000'))->getDetails();
        $this->expectException(\LogicException::class);
        $details->getArea();
    }

    /**
     * @dataProvider validityPhoneNumbers
     */
    public function test_can_get_local_format(string $number)
    {
        $localFormat = (new PhoneNumber($number))->toLocalFormat();
        $this->assertEquals('0710000000', $localFormat);
    }

    /**
     * @dataProvider validityPhoneNumbers
     */
    public function test_can_get_IDD_format(string $number)
    {
        $localFormat = (new PhoneNumber($number))->toIDDFormat();
        $this->assertEquals('+94710000000', $localFormat);
    }

    public function validityPhoneNumbers(): array
    {
        return [
            ['0710000000'],
            ['+94710000000'],
            ['0094710000000'],
        ];
    }
}


src/PhoneNumberDetails.php
<?php

namespace APiChef\SlPhoneNumber;

class PhoneNumberDetails
{
    /** @var string */
    private $number;

    /** @var string */
    private $type;

    /** @var string */
    private $operator;

    /** @var string */
    private $province = null;

    /** @var string */
    private $district = null;

    /** @var string */
    private $area = null;

    public function __construct(string $number, string $type, string $operator, array $location = null)
    {
        $this->number = $number;
        $this->type = $type;
        $this->operator = $operator;

        if ($location !== null) {
            $this->province = $location['province'];
            $this->district = $location['district'];
            $this->area = $location['area'];
        }
    }

    public function getNumber(): string
    {
        return $this->number;
    }

    public function getType(): string
    {
        return $this->type;
    }

    public function getOperator(): string
    {
        return $this->operator;
    }

    public function getProvince(): string
    {
        $this->checkLocationDetails();

        return $this->province;
    }

    public function getDistrict(): string
    {
        $this->checkLocationDetails();

        return $this->district;
    }

    public function getArea(): string
    {
        $this->checkLocationDetails();

        return $this->area;
    }

    private function checkLocationDetails(): void
    {
        if ($this->province === null) {
            throw new \LogicException('Can not read location details form mobile number');
        }
    }
}


src/PhoneNumber.php
<?php

namespace APiChef\SlPhoneNumber;

class PhoneNumber
{
    const TYPE_MOBILE = 'mobile';
    const TYPE_FIXED = 'fixed';

    private $areaCodes = [
        '011' => ['province' => 'Western', 'district' => 'Colombo', 'area' => 'Colombo'],
        '021' => ['province' => 'Northern', 'district' => 'Jaffna', 'area' => 'Jaffna'],
        '023' => ['province' => 'Northern', 'district' => 'Mannar', 'area' => 'Mannar'],
        '024' => ['province' => 'Northern', 'district' => 'Vavuniya', 'area' => 'Vavuniya'],
        '025' => ['province' => 'North Central', 'district' => 'Anuradhapura', 'area' => 'Anuradhapura'],
        '026' => ['province' => 'Eastern', 'district' => 'Trincomalee', 'area' => 'Trincomalee'],
        '027' => ['province' => 'North Central', 'district' => 'Polonnaruwa', 'area' => 'Polonnaruwa'],
        '031' => ['province' => 'Western', 'district' => 'Gampaha', 'area' => 'Negombo'],
        '032' => ['province' => 'North Western', 'district' => 'Puttalam', 'area' => 'Chilaw'],
        '033' => ['province' => 'Western', 'district' => 'Gampaha', 'area' => 'Gampaha'],
        '034' => ['province' => 'Western', 'district' => 'Kalutara', 'area' => 'Kalutara'],
        '035' => ['province' => 'Sabaragamuwa', 'district' => 'Kegalle', 'area' => 'Kegalle'],
        '036' => [
            'province' => 'Western,Sabaragamuwa',
            'district' => 'Colombo,Rathnapura,Kegalle',
            'area' => 'Avissawella'
        ],
        '037' => ['province' => 'North Western', 'district' => 'Kurunegala', 'area' => 'Kurunegala'],
        '038' => ['province' => 'Western', 'district' => 'Kalutara', 'area' => 'Panadura'],
        '041' => ['province' => 'Southern', 'district' => 'Matara', 'area' => 'Matara'],
        '045' => ['province' => 'Sabaragamuwa', 'district' => 'Ratnapura', 'area' => 'Ratnapura'],
        '047' => ['province' => 'Southern', 'district' => 'Hambantota', 'area' => 'Hambantota'],
        '051' => ['province' => 'Central', 'district' => 'Nuwara Eliya', 'area' => 'Hatton'],
        '052' => ['province' => 'Central', 'district' => 'Nuwara Eliya', 'area' => 'Nuwara Eliya'],
        '054' => ['province' => 'Central', 'district' => 'Kandy', 'area' => 'Nawalapitiya'],
        '055' => ['province' => 'Uva', 'district' => 'Monaragala,Badulla', 'area' => 'Monaragala,Badulla'],
        '057' => ['province' => 'Uva', 'district' => 'Badulla', 'area' => 'Bandarawela'],
        '063' => ['province' => 'Eastern', 'district' => 'Ampara', 'area' => 'Ampara'],
        '065' => ['province' => 'Eastern', 'district' => 'Batticaloa', 'area' => 'Batticaloa'],
        '066' => ['province' => 'Central', 'district' => 'Matale', 'area' => 'Matale'],
        '067' => ['province' => 'Eastern', 'district' => 'Ampara', 'area' => 'Kalmunai'],
        '081' => ['province' => 'Central', 'district' => 'Kandy', 'area' => 'Kandy'],
        '091' => ['province' => 'Southern', 'district' => 'Galle', 'area' => 'Galle'],
    ];

    private $operatorCodes = [
        '0' => ['name' => 'Lanka Bell', 'type' => 'Fixed LTE'],
        '2' => ['name' => 'Sri Lanka Telecom', 'type' => 'Fixed Fibre or Copper'],
        '3' => ['name' => 'Sri Lanka Telecom', 'type' => 'Fixed CDMA or LTE'],
        '4' => ['name' => 'Dialog', 'type' => 'Fixed LTE'],
        '5' => ['name' => 'Lanka Bell', 'type' => 'Fixed CDMA'],
        '7' => ['name' => 'Dialog', 'type' => 'Fixed LTE'],
        '9' => ['name' => 'Tritel', 'type' => 'Public Payphones'],
    ];

    private $mobileOperatorCodes = [
        '070' => ['name' => 'Mobitel'],
        '071' => ['name' => 'Mobitel'],
        '072' => ['name' => 'Hutch'],
        '074' => ['name' => 'Dialog'],
        '075' => ['name' => 'Airtel'],
        '076' => ['name' => 'Dialog'],
        '077' => ['name' => 'Dialog'],
        '078' => ['name' => 'Hutch'],
    ];

    /** @var string */
    private $number;

    public function __construct(string $number)
    {
        $this->number = $number;
    }

    public function getDetails(): PhoneNumberDetails
    {
        if ($this->isFormatValid()) {
            $number = $this->toLocalFormat();
            $prefix  = substr($number, 0, 3);

            if (array_key_exists($prefix, $this->mobileOperatorCodes)) {
                return new PhoneNumberDetails($number, self::TYPE_MOBILE, $this->mobileOperatorCodes[$prefix]['name']);
            }

            if (array_key_exists($prefix, $this->areaCodes)) {
                $operatorCode = substr($number, 3, 1);
                if (array_key_exists($operatorCode, $this->operatorCodes)) {
                    return new PhoneNumberDetails(
                        $number,
                        self::TYPE_FIXED,
                        $this->operatorCodes[$operatorCode]['name'],
                        $this->areaCodes[$prefix]
                    );
                }
            }
        }

        throw new \InvalidArgumentException('Invalid phone number.');
    }

    public function isValid(): bool
    {
        try {
            $this->getDetails();

            return true;
        } catch (\InvalidArgumentException $exception) {
            return false;
        }
    }

    private function isFormatValid(): bool
    {
        return preg_match('/^0[0-9]{9}$/', $this->number) ||
            preg_match('/^\+94[0-9]{9}$/', $this->number) ||
            preg_match('/^0094[0-9]{9}$/', $this->number);
    }

    public function toLocalFormat(): string
    {
        $number = $this->number;

        if (substr($this->number, 0, 3) === '+94') {
            $number = '0' . substr($this->number, 3);
        }

        if (substr($this->number, 0, 4) === '0094') {
            $number = '0' . substr($this->number, 4);
        }

        return $number;
    }

    public function toIDDFormat(): string
    {
        return '+94' . substr($this->toLocalFormat(), 1);
    }
}
